dns-server/src/dns_server.rs

```diff
@@ -22,191 +31,244 @@ use trust_dns_proto::rr::{Name, Record};
+                        rb_servfail,
+                        &header,
+                    )
+                    .await
+                }
+                RequestError::ServFail(_) => {
+                    respond_servfail(&request, rb_servfail, &header).await
+                }
+            };
+        }
     }
+}
 
-    let name = mr.query().original().name().clone();
-    let key = name.to_string();
-    let key = key.trim_end_matches('.');
-
-    let rb = MessageResponseBuilder::from_message_request(&mr);
-
-    let bits = match db.get(key.as_bytes()) {
-        Ok(Some(bits)) => bits,
-
-        // If no record is found bail with NXDOMAIN.
-        Ok(None) => {
-            respond_nxdomain(&log, socket, src, rb, header, &mr).await;
-            return;
-        }
+/// Describes how to respond to a particular request failure
+#[derive(Debug, Error)]
+enum RequestError {
+    #[error("NXDOMAIN: {0:#}")]
+    NxDomain(#[source] QueryError),
+    #[error("SERVFAIL: {0:#}")]
+    ServFail(#[source] anyhow::Error),
+}
 
-        // If we encountered an error bail with SERVFAIL.
-        Err(e) => {
-            error!(log, "db get: {}", e);
-            nack(&log, &mr, &socket, &header, &src).await;
-            return;
+impl From<QueryError> for RequestError {
+    fn from(source: QueryError) -> Self {
+        match &source {
+            QueryError::NoName(_) => RequestError::NxDomain(source),
+            // Bail with servfail when this query is for a zone that we don't
+            // own (and other server-side failures) so that resolvers will look
+            // to other DNS servers for this query.
+            QueryError::NoZone(_)
+            | QueryError::QueryFail(_)
+            | QueryError::ParseFail(_) => RequestError::ServFail(source.into()),
         }
-    };
-
-    let records: Vec<crate::dns_data::DnsRecord> =
-        match serde_json::from_slice(bits.as_ref()) {
-            Ok(r) => r,
-            Err(e) => {
-                error!(log, "deserialize record: {}", e);
-                nack(&log, &mr, &socket, &header, &src).await;
-                return;
-            }
-        };
-
-    if records.is_empty() {
-        error!(log, "No records found for {}", key);
-        respond_nxdomain(&log, socket, src, rb, header, &mr).await;
-        return;
     }
+}
+
+/// Handle a well-formed, decoded DNS query
+async fn handle_dns_message(
+    request: &Request,
+    mr: &MessageRequest,
+) -> Result<(), RequestError> {
+    let log = &request.log;
+    let store = &request.store;
+    debug!(&log, "message_request"; "mr" => #?mr);
 
-    let mut response_records: Vec<Record> = vec![];
-    for record in &records {
-        let resp = match record {
+    let header = Header::response_from_request(mr.header());
+    let query = mr.query();
+    let name = query.original().name().clone();
+    let records = store.query(mr)?;
+    let rb = MessageResponseBuilder::from_message_request(mr);
+    let response_records = records
```

Comment:
If this result comes back empty, we should probably check if the logical DNS tree rooted at the queried name is also empty, and if so return an NXDOMAIN per RFCs 1034 and 8020. I could see this happening a few different ways.

1. If we stored the keys (names) in reverse, we could use sled's `scan_prefix` under the hood to do a prefix check.
2. We look at the query and see if the name extends beyond the zone prefix, and if so, just assume this is a terminal name and that no names exist below it. I think this may work with the constrained way we are using DNS, but I could be missing something.

Reply:
Good call.  I believe this case is handled already, if somewhat implicitly.  The contract of `query()` is that it returns a non-empty list of records or else an error.  This wasn't documented, though.  I've documented it in 2ffdb9e8.

More specifically: if the name were not found in the tree at all, `query()` would return `QueryError::NoName` here:
https://github.com/oxidecomputer/omicron/blob/e86747334326bb001eb6f55f731357a47ee4f96c/dns-server/src/storage.rs#L636

The storage implementation never puts a name in the tree with an empty list of records, but it handles the case where it finds that by returning `QueryError::NoName` instead:
https://github.com/oxidecomputer/omicron/blob/e86747334326bb001eb6f55f731357a47ee4f96c/dns-server/src/storage.rs#L644-L656

So we should always get `QueryError::NoName` for this case, and that gets translated to NXDOMAIN:
https://github.com/oxidecomputer/omicron/blob/2ffdb9e8a222e2a43aeb8a62a958ab6088c96d96/dns-server/src/dns_server.rs#L206

I think this corresponds to your suggestion 2.  Thanks for mentioning it though because this was rather implicit.  (I could make it even more explicit by having `query()` return not a `Result`, but an enum that includes the various errors plus a `NonEmptyRecords(Vec<DnsRecord>)` variant.  But that's more unwieldy -- you can't use `?` to return early and convert it to the DNS error, for example.)

Reply:
Is that notation in reversed order?  For the external endpoints, I imagine the customer will delegate us something like `oxide.corp.globex.example`, they'll create Silo `prod` (say), and we'll have records for `prod.sys.oxide.corp.globex.example`.  There won't be anything at `sys.oxide.corp.globex.example`.  There will presumably be NS records for `oxide.corp.globex.example`.  I suspect this violates the assumption you said.  I'm not sure what the impact is?  What's the right DNS response for `sys.oxide.corp.globex.example`?  A normal response with zero records?

As an aside: I think the behavior here hasn't changed in this PR.  (Is that wrong?)

Reply:
Yes, this is in reverse order. Sorry, my brain was still thinking in DNS names as trees, and I was starting from the root.

So if we have no records at `sys.oxide.corp.globex.example` but a record logically below that in the tree `prod.sys.oxide.corp.globex.example` exists, the correct response for `sys.oxide.corp.globex.example` would be a normal response with no records.

I probably got this wrong in my initial implementation too.

Reply:
SGTM